### DISCUSSION TITLE
running docker compose in hetzner cloud with non default ports possible?

### DISCUSSION THREAD

**Discussion by sgbx** [kimai/kimai#5201] [Docker] [answered]
### Describe the problem

I tried to install kimai following the instructions from https://www.kimai.org/documentation/hosting-hetzner-cloud.html
Unfortunately, the ports 443 and 80 are already taken by other services on my server. Is there a way to make it run with different setup?
I tried 
`services:
  caddy:
    image: caddy:latest
    restart: unless-stopped
    ports:
      - "81:80"
      - "444:443"
`
but it did not work.

### Describe your setup and add your Docker compose file (redact your credentials)

debian bookworm
docker compose file 3.7

### Command used to run the container

docker compose up -d

**Comment by kevinpapst** [comment]
Your question doesn't make a lot of sense.
The documentation is for a new server without other services.

If you want to run Kimai next to your existing setup, you don't need Caddy as reverse proxy.
You can "simply" plug Kimai into your existing webserver / proxy.

**Reply by sgbx** [reply]
A OK

> Your question doesn't make a lot of sense. The documentation is for a new server without other services.
> 
> If you want to run Kimai next to your existing setup, you don't need Caddy as reverse proxy. You can "simply" plug Kimai into your existing webserver / proxy.

Thank you, I see. So I would follow https://www.kimai.org/documentation/docker-compose.html and set ports e.g. to 4444 in docker-compose.yml where a reverse proxy would be listening to internally?

**Reply by kevinpapst** [reply, marked as answer]
Use the apache image and adapt it to your needs, yes.
You can re-use existing database. And you need to adapt 
```
    ports:
      - 8001:8001
```
to fit into your environment. The first element needs to be changed to whatever your reverse proxy listens to.

**Comment by sgbx** [comment]
That all worked well, @kevinpapst 
However, I am now unable to log in with my email address and password stored in the .env. Is there a different user that I have to use the first time? I haven't found anything about this on google, in the documentation or in the discussions here.

**Reply by kevinpapst** [reply]
Any errors in the Docker logs?
The Admin User should be auto-created during image boot.

**Reply by sgbx** [reply]
Maybe this looks strange
 [OK] Successfully installed Kimai version 2.25.0 🎉                            

+ '[' '!' -z mypasswordi ']'
+ '[' '!' -a mymailadress ']'
+ /opt/kimai/bin/console kimai:user:create admin mymailadress ROLE_SUPER_ADMIN mypassword

 [ERROR] email: This e-mail address is already in use.                          

 [ERROR] username: The username is already used.

**Reply by kevinpapst** [reply]
That means the admin account exists and you should be able to login with it

**Reply by sgbx** [reply]
Any idea, what could go wrong here?
I put also TRUSTED_PROXIES in the yaml and checked permissions from the trouble shooting page. 
Maybe something wrong in the proxy settings?
    location / {
          proxy_pass http://127.0.0.1:4444;
          proxy_set_header  Host $http_host;
          proxy_set_header  X-Real-IP $remote_addr;
          proxy_set_header  X-Forwarded-Host $host:$server_port;
          proxy_set_header  X-Forwarded-Server $host;
          proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header  X-Forwarded-Proto $scheme;
          proxy_set_header  X-Forwarded-Port $server_port;
    }

**Reply by kevinpapst** [reply]
> Any idea, what could go wrong here?

You don't even share "what" goes wrong.
Are you referring to

> However, I am now unable to log in with my email address and password stored in the .env. 

In the .env is no username and password. You have to use these for the login:
```
ADMINMAIL - Email for the new admin user (will be created if not existing)
ADMINPASS - Password for the new admin user (will be created if not existing)
```

And then"login doesn't work" doesn't explain much. Do you see the Kimai login UI? Is there an error?

**Reply by sgbx** [reply]
Ok, I'm sorry. 
I did set my personal adminmail and adminpass in the .env. I can see the kimai login UI under my domain adress after I run docker compose up -d
When I try to log in with my adminmail and my adminpass and hit the enter button, my firefox browser gives me a warning (HTTPS-Only Mode Alert
Secure Site Not Available
You’ve enabled HTTPS-Only Mode for enhanced security, and a HTTPS version of kimai.pagenstedt.de is not available.)
My domain has hsts enabled, but I never had an issue like this with other self hosted services. Of course, I have a proper ssl cerfiticate and cleared the browser caches.

**Reply by kevinpapst** [reply]
Apache image? This might be an issue in the Kimai image, where Kimai doesn't see the proxy, becuase Apache doesn't forward the headers. There is a possible fix, but nobody was yet able to verify it: https://github.com/kimai/kimai/pull/5071

**Reply by kevinpapst** [reply]
But wait. After re-reading this one: I think your issue is different. 
Your reverse proxy serves a HTTP site on port 443. 
That likely means you did not yet configure SSL for the Kimai vhost.
You have to serve and terminate SSL on your proxy. 
The other issue is about IP logging, so not related.

**Reply by sgbx** [reply]
> But wait. After re-reading this one: I think your issue is different. Your reverse proxy serves a HTTP site on port 443. That likely means you did not yet configure SSL for the Kimai vhost. You have to serve and terminate SSL on your proxy. The other issue is about IP logging, so not related.

Thank you. What would I have to change here exactely? I do have an certificate vor my kimai domain and the site can be reached via https. 
I also tried the APP_URL with https in docker-compose.yaml, but with no effect.
How do I have to configure SSL for the Kimai vhost, another certificate for the docker apache?

**Reply by sgbx** [reply]
Hey @kevinpapst - just wanted to let you know, everything runs fine now. The problem was the internal port 4444. It seems, my provider Hetzner does not like this port. I read this port is often used by malware, so maybe a filter applied. 
Thanks again for your support and the nice peace of software.
